### DISCUSSION TITLE
How should I fine-tune Qwen3-4B-Base on raw novel text without a chat/QA format?

### DISCUSSION THREAD

**Discussion by kbuwel** [QwenLM/Qwen3#1908] [Q&A]
Hello,

I am trying to fine-tune **Qwen3.5-4B-Base** on Korean novel text using QLoRA.

My goal is **not instruction tuning, chatbot SFT, roleplay training, or question/answer training**.

I want the model to learn from raw novel text for:

- natural continuation of prose
- writing style
- narrative patterns
- characters/settings/context contained in the corpus

In other words, what I am trying to do is closer to **continued pretraining / causal language modeling on a small domain corpus**, but using QLoRA because I have limited VRAM.

### Current dataset format

My dataset is stored in Parquet.

Each row contains one complete novel/story in a `text` column.

I do not convert the text into:

```text
user:
assistant:
```

or any instruction/chat template.

Before training, I tokenize all stories as raw text and concatenate them into a continuous causal-LM token stream.

I currently use:

- chapter separator: `***`
- book separator: `⁂`
- sequence length: 2048 tokens
- fixed-length packing
- about 10% of packed blocks for validation

The structure is roughly:

```text
Novel 1 text...
***
next chapter...
⁂
Novel 2 text...
***
next chapter...
⁂
Novel 3 text...
```

This token stream is then divided into 2048-token blocks.

I do not truncate each novel to 2048 tokens. The complete text is tokenized and distributed across multiple blocks.

### Causal LM labels

For training, the labels are simply a copy of `input_ids`.

Only actual padding positions are masked with `-100`.

Conceptually:

```python
labels = input_ids.clone()
labels[attention_mask == 0] = -100
```

I do this because the tokenizer may use the same ID for PAD and EOS, and I do not want real EOS tokens to be accidentally excluded from the loss.

### QLoRA configuration

I load the model using 4-bit NF4 quantization:

```python
BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=torch.bfloat16
        if torch.cuda.is_bf16_supported()
        else torch.float16,
)
```

My current LoRA configuration is:

```python
LoraConfig(
    r=32,
    lora_alpha=64,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=lora_target_modules,
)
```

I try to include both attention and MLP projections.

For the Qwen3.5 hybrid architecture, my code detects available modules from the model and includes modules such as:

```text
q_proj
k_proj
v_proj
o_proj

in_proj_qkv
in_proj_z
in_proj_b
in_proj_a
out_proj

gate_proj
up_proj
down_proj
```

depending on which modules actually exist in the loaded model.

### Training settings

The main settings are approximately:

```text
per_device_train_batch_size = 1
max_seq_length = 2048
optimizer = paged_adamw_8bit
lr_scheduler = cosine
weight_decay = 0.01
gradient_checkpointing = True
```

The script automatically selects gradient accumulation, learning rate and number of optimizer updates according to corpus size.

For a corpus up to about 1M tokens, for example, it currently chooses approximately:

```text
optimizer updates = 256
gradient accumulation = 4
learning rate = 7e-5
warmup ratio = 0.03
```

### The problem

Training itself works, and the loss decreases.

However, the resulting model does not seem to learn the novel corpus in the way I expected.

In generation tests, I sometimes see:

- unrelated topics/settings appearing
- the model drifting toward concepts that do not fit the source context
- weak preservation of the narrative context
- outputs that seem dominated by the base model's prior knowledge rather than the novel corpus
- in some experiments, multiple unrelated prompts drifting toward similar topics

Increasing or decreasing LoRA rank alone has not clearly solved the problem.

Because of this, I am no longer sure whether the problem is:

1. my dataset construction,
2. my packing strategy,
3. the number of training updates,
4. the learning rate,
5. my LoRA target modules,
6. QLoRA itself,
7. or whether this task should be treated differently from normal LoRA SFT.

### My questions

I would especially appreciate advice on the following:

**1. Is raw causal-LM training like this appropriate for Qwen3.5-4B-Base?**

For novel continuation/style learning, should I simply train on raw token sequences with:

```text
labels = input_ids
```

without using a chat template?

**2. Is this better considered continued pretraining rather than SFT?**

If so, is PEFT/LoRA suitable for this, or is full-parameter continued pretraining normally required to get meaningful results?

**3. Is concatenating multiple novels into one packed token stream reasonable?**

I currently insert a separator between books before packing.

Would it be better to reset sequences at document boundaries instead of allowing a 2048-token block to contain the end of one book and the beginning of another?

**4. Are my LoRA targets appropriate for Qwen3.5?**

Should the Gated DeltaNet / linear-attention projections and MLP projections be trained for this task, or would targeting fewer modules work better?

**5. Is `r=32, alpha=64` reasonable for this type of domain adaptation?**

Would a larger rank such as 64 actually help with prose/domain learning, or is data quantity/training duration likely to be much more important?

**6. Is approximately 256 optimizer updates far too little for continued pretraining?**

For a relatively small novel corpus, should I think in terms of epochs/tokens seen rather than a fixed number of optimizer steps?

**7. What is the recommended way to fine-tune a Qwen Base model specifically for raw text continuation?** 
Most fine-tuning examples I find focus on instruction/chat datasets. I would like to know the recommended approach for plain text such as books, articles, or domain-specific corpora.

I can provide the complete training script and training/generation logs if they would be useful.

Thank you.

**Comment by Humerous** [comment]
Yes — for the objective you describe, I would treat this as continued pretraining / domain-adaptive causal language modeling rather than instruction SFT.

You do not need a chat template for this. For a Base model, training directly on raw prose with next-token prediction is the appropriate objective:

```python
labels = input_ids.clone()
labels[attention_mask == 0] = -100

**Comment by helloe365** [comment]
Agreed with the comment above on the core point — this is continued pretraining, not SFT, and raw next-token prediction is exactly right for a base model. Chat templates would actively hurt here.

On the rest: the symptoms you describe (generations dominated by base-model priors, unrelated prompts converging to similar outputs) are the classic signature of an undertrained adapter, not a broken pipeline. Some specifics:

**256 optimizer updates is way too few.** With batch 1 and grad accumulation 4 you're seeing ~8k tokens per step, so 256 steps is roughly 2 epochs over 1M tokens. Style/domain adaptation typically needs the corpus seen 20-50x (think in tokens-seen rather than steps: ~20-50M tokens for your size). This alone probably explains most of what you're observing — I'd fix this before touching anything else.

**lr 7e-5 is low for QLoRA.** 1e-4 to 2e-4 is the usual band for this kind of run; with the longer schedule the cosine decay actually gets somewhere.

**r=32 is fine to start.** Capacity is rarely the binding constraint at 1M tokens — try 64/128 only if a much longer run still underfits. Keeping the DeltaNet projections in the target list is reasonable too, from LoRA's perspective they're just more projections.

**Packing is fine as-is** — concatenating whole books with separators is standard practice for prose corpora. The one change I'd make: end each novel with the model's actual EOS token in addition to your ⁂ separator, so it gets a chance to learn "story ends here" instead of "stories continue forever". The cross-document attention inside a 2048 block is a second-order effect compared to simply being undertrained.

**Validate with held-out perplexity**, not by eyeballing generations. If val loss is still falling steeply when you stop, you stopped early — your 10% split will tell you this immediately.

Full-parameter training becomes interesting with 100M+ tokens or when you need to unlearn priors; neither applies here, so QLoRA is a reasonable fit. And there's no framework needed for this — your own Trainer setup is the right tool; most chat-oriented frameworks assume instruction data anyway.

Concrete next run: same script, ~20-30 epochs, lr 1e-4, eval every few epochs, early-stop on val loss. My guess is the drifting disappears well before epoch 20.

If you post the training logs, the loss curve should confirm the undertrained theory at a glance — and if this ends up solving it, marking it as the answer would help the next person attempting raw-text CPT.
